Ensure that the socket is removed when api.SetDefaultPermissions fail

kaworu opened the pull request:
Consistently remove the socket when we fail to set its owner and/or permission.

rolinh reviewed (approved):
LGTM for the Hubble changes.

aanm commented on api/v1/server.gotmpl:
```diff
@@ -331,6 +331,7 @@ func (s *Server) Serve() (err error) {
 		if os.Getuid() == 0 {
 		    err := api.SetDefaultPermissions(string(s.SocketPath))
 			if err != nil {
+				os.Remove(string(s.SocketPath))
```
`s.SocketPath` is already a string.

aanm reviewed (changes requested):
Just one small thing, what happens if there's an instance already running, can we somehow accidentally delete the socket path?

gandro reviewed (approved):
Looks good to me! (besides the nit that André found)

sayboras commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -72,6 +72,8 @@ func WithUnixSocketListener(path string) Option {
 		}
 		if os.Getuid() == 0 {
 			if err := api.SetDefaultPermissions(socketPath); err != nil {
+				socket.Close()
```
I think socket.Close() will calll syscall.Unlink internally if I am not wrong ?

gandro replied:
You are correct! https://github.com/golang/go/blob/a38a917aee626a9b9d5ce2b93964f586bf759ea0/src/net/unixsock_posix.go#L169-L187

I guess we could set `SetUnlinkOnClose` and remove the unlink and just have a close?

sayboras commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -72,6 +72,8 @@ func WithUnixSocketListener(path string) Option {
 		if os.Getuid() == 0 {
 			if err := api.SetDefaultPermissions(socketPath); err != nil {
+				socket.Close()
+				unix.Unlink(socketPath)
```
another question for my understanding. Any reason to use `golang.org/x/sys/unix` instead of syscall.Unlink ? Thanks.

rolinh replied:
See #10116 for this one.

sayboras replied:
Thanks for your info 💯

I was under impression that x package is kind of experiment only, recently just replaced xerrors package at work.

kaworu committed:
api/v1: use net.UnixListener for unix sockets

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

kaworu committed:
hubble/server: close the socket when api.SetDefaultPermissions fail

Also fail fast when o.Listener is already configured as it is a
programming or configuration issue.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

kaworu committed:
monitor: use net.UnixListener for unix sockets

Also ensure that the socket is closed and file removed on failure.

Signed-off-by: Alexandre Perrin <alex@kaworu.ch>

sayboras commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -64,22 +64,28 @@ func WithTCPListener(address string) Option {
 // owner is set to socketGroup.
 func WithUnixSocketListener(path string) Option {
 	return func(o *Options) error {
+		if o.UnixSocketListener != nil {
```
do we need to call o.UnixSocketListener.Close() in this if condition ?

kaworu replied:
Looking only at this function, we either return an error leaving `o` in the same state we found it or we modify `o` with success. This is a clean behaviour and closing an already configured socket would break this promise.

As it is currently used, when this codepath is hit `pkg/hubble/server/server.NewServer` (i.e. the caller) would return a "failed to apply option" error and we'll loose ref to the `Options` struct (and thus the socket).

I would say we leave it as-is because it's corner case enough to let the garbage collector close the socket at some point. Alternatively, I would be in favour of a `func (o *Options) Cleanup()` to be called by `pkg/hubble/server/server.NewServer` on failure.

sayboras replied:
thanks for your explanation, it's making sense to me :tada:

sayboras reviewed (approved):
One small question.

I learned new stuffs in this PR :100:

rolinh commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -34,7 +34,7 @@ import (
 type Options struct {
 	TCPListener        net.Listener
-	UnixSocketListener net.Listener
+	UnixSocketListener *net.UnixListener
```
PR #12906 (still draft) will remove support for multiple listeners as this is a feature that is unused. However, this means that `WithUnixSocketListener()` and `WithTCPListener()`, which are mutually exclusive options, both set a single listener which has to be of type `net.Listener`. Can we somehow keep `net.Listener` as the type here? I'll have to revert that change anyway when rebasing #12906 against master with the changes in this PR.

kaworu replied:
@Rolinh we'll block this PR waiting on yours as this one is simpler and less impactful. If we end up with a single listener that is not necessarily a socket, then it has to be a `net.Listener` of course.

kaworu replied:
Kept `net.Listener` as discussed when rebased.

sayboras reviewed (approved):
LGTM :100:
